### ISSUE TITLE
Newly created objects lack parent object links.

### ISSUE THREAD

**Issue by JoeThielen** [graphaware/neo4j-php-ogm#103]
I might mess up the terminology here so please bear with me.

If I create 3 new nodes via neo4j-php-ogm (RC1 & dev-master, tried both specifically) with simple relationships like X->Y->Z, then immediately do a findOneBy() on Y and do a print_r() on it, the arrayCollection for Z is OK and correct, but the link (not a collection since there can only be one in this use case) back to X does not populate.  Doing something like a Y->getX() results in null, whereas Y->getZs() is OK.  

It does populate the DB correctly.  And if, instead of immediately doing a findOneBy(), I first destroy then create a new EntityManager, then do the findOneBy(), then it works perfect.

It seems partially similar to #21 , at least, in terms of odd behavior with new objects.

I can create a full set of test data if need be, but I was hoping the above would be clear enough.

**Comment by JoeThielen** [comment]
I tried to create a demo script using the play Movies data, but that actually worked and did not fail as expected.  However, in the Movies sample both Movie and Person have many-to-many relationships to each other.  In my use case, in the X->Y->Z example, Y would only link back to one X, so it's not a collection.  And my code works perfect when I refresh the entityManager, which I assume forces it to read back from the DB, so I think I've got my mappings/annotations functioning correctly.

**Comment by JoeThielen** [comment]
OK I've gone ahead and created test scripts which demonstrate the issue I am having.

Entity.php:
```
<?php

use Doctrine\Common\Collections\ArrayCollection;
use GraphAware\Neo4j\OGM\Annotations as OGM;

/**
 * @OGM\Node(label="Entity")
 */
class Entity
{
    /**
     * @OGM\GraphId()
     * @var int
     */
    protected $id;

    /**
     * @OGM\Property(type="string")
     * @var string
     */
    protected $name;


    /**
     * @OGM\Relationship(type="HAS_CONTEXT", direction="OUTGOING", targetEntity="\Context", collection=true, mappedBy="entity")
     * @var ArrayCollection|\Context[];
     */
    protected $contexts;

    public function __construct($name)
    {
        $this->name = $name;

        $this->contexts = new ArrayCollection();
    }

    /**
     * @return int
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     * @return string
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * @param string $name
     */
    public function setName($name)
    {
        $this->name = $name;
    }

    /**
     * @return \Doctrine\Common\Collections\ArrayCollection|\Context[];
     */
    public function getContexts()
    {
        return $this->contexts;
    }

    /**
     * @param \Context $context
     */
    public function addContext(\Context $context)
    {
        if (!$this->contexts->contains($context)) {
            $this->contexts->add($context);
        }
    }

    /**
     * @param \Context $context
     */
    public function removeContext(\Context $context)
    {
        if ($this->contexts->contains($context)) {
            $this->contexts->removeElement($context);
        }
    }
}
```

Context.php:
```
<?php

use Doctrine\Common\Collections\ArrayCollection;
use GraphAware\Neo4j\OGM\Annotations as OGM;

/**
 * @OGM\Node(label="Context")
 */
class Context
{
    /**
     * @OGM\GraphId()
     * @var int
     */
    protected $id;

    /**
     * @OGM\Property(type="string")
     * @var string
     */
    protected $name;

    /**
     * @OGM\Relationship(type="HAS_CONTEXT", direction="INCOMING", targetEntity="Entity", collection=false, mappedBy="contexts")
     * @var string
     */
    protected $entity;

    /**
     * @param string $name
     */
    public function __construct($name)
    {
        $this->name = $name;
    }

    /**
     * @return int
     */
    public function getId()
    {
        return $this->id;
    }

    /**
     * @return string
     */
    public function getName()
    {
        return $this->name;
    }

    /**
     * @param string $name
     */
    public function setName($name)
    {
        $this->name = $name;
    }

    /**
     * @return string
     */
    public function getEntity()
    {
        return $this->entity;
    }
}
```

test.php:
```
<?php

require_once __DIR__ . "/vendor/autoload.php";
require_once __DIR__ . "/Entity.php";
require_once __DIR__ . "/Context.php";

use GraphAware\Neo4j\OGM\EntityManager;


// CREATE ENTITY MANAGER
$manager = EntityManager::create('http://localhost:7474');


// CREATE ENTITY
$test_entityUuid = uuid_create(4);
$entity = new Entity($test_entityUuid);


// CREATE CONTEXT
$test_contextUuid = uuid_create(4);
$context = new Context($test_contextUuid);


// ADD CONTEXT TO ENTITY
$entity->addContext($context);
echo "REFERENCE ENTITY UUID:|" . $test_entityUuid . "|\n";


// SAVE EVERYTHING
$manager->persist($entity);
$manager->persist($context);
$manager->flush();


// LOOK UP CONTEXT
$context2 = $manager->getRepository(Context::class)->findOneBy(array('name' => $test_contextUuid));

// THIS WILL NEED TO BE COMMENTED OUT IN ORDER TO MOVE ON TO TRY #2!!!
echo "TRY #1: |" . print_r($context2->getEntity()->getName(), 1) . "|\n";


// REFRESH ENTITY MANAGER
$manager = EntityManager::create('http://localhost:7474');


// LOOK UP CONTEXT
$context2 = $manager->getRepository(Context::class)->findOneBy(array('name' => $test_contextUuid));
echo "TRY #2: |" . print_r($context2->getEntity()->getName(), 1) . "|\n";
```

Process:
- Put above 3 files into directory
- Add composer.json with requirement for neo4j-php-ogm and run composer install
- Run test.php, it should error on TRY 1 with this error:
> PHP Fatal error:  Uncaught Error: Call to a member function getName() on null in /var/www/html/xtest2/test.php:39
- Edit test.php, comment out the TRY 1 echo line.
- Run test.php, it should now succeed with TRY 2, which only runs after refreshing the entity manager.

I'm still new to this so maybe I'm doing something wrong here.  But I don't get that it works OK after refreshing the entity manager.

**Comment by JoeThielen** [comment]
After re-examining my examples, I see I am lacking a setter for Context->$entity.  I went ahead and added it, but it still did not self-populate.  If, in test script, I manually did it, then of course it did work.  For example:

If I changed:
```
// ADD CONTEXT TO ENTITY
$entity->addContext($context);
echo "REFERENCE ENTITY UUID:|" . $test_entityUuid . "|\n";
```

and added:
```
// ADD CONTEXT TO ENTITY
$entity->addContext($context);
$context->setEntity($entity);
echo "REFERENCE ENTITY UUID:|" . $test_entityUuid . "|\n";
```

Then it works.  But why would I have to create the same relationship twice?

Anyway, so this works, **EXCEPT** it becomes a **HUGE** memory hog!  In the X->Y model in my example it's not too bad (but still not great), but when I start doing X->Y->Z I can easily eat up 1GB of memory in my unit tests!  Wow.

For the moment I've found that the simple answer is to not rely on the EntityManager after the initial creation.   I instead destroy the EntityManager after flush()ing, then re-instantiate it, which then forces it to load from the DB, and at that point it all works.  No memory issues there.  But now I'm wondering about possible performance issues.

**Comment by ikwattro** [comment]
Thanks for this detailed report @JoeThielen , looking at it now

**Comment by ikwattro** [comment]
@JoeThielen 

Regarding the fact that the entity object is not available on your context object has nothing to do with the ogm but with your object graph. If you wouldn't use any database the behavior would be the same. 

So now comes some exceptions that I wanted to introduce but I wanted to have a bit more community feedback before it. Because you don't have a bi-directional reference in your object graph and in neo4j you cannot have orphan relationships (which is what you have here before trying to flush) should the ogm throw an exception instead of saving the graph optimistically. 

This is something I will think further about.

To summarize, you are not creating the relationship twice by setting the entity on the context, you are in fact enforcing the fact of having a consistent object graph.

Regarding performance issues, I didn't encounter them but I will anyway make sure to have a 3 depths graph in the tests to cover performances.

**Comment by JoeThielen** [comment]
I see what you are saying.  I am not sure if the docs made that clear, which I think may have been the source of my confusion.  If you are going to enforce this behavior then I would highly suggest ensuring the docs make this point extremely clear for other graph newbies such as myself!

It's very confusing in the fact that it "just works" when pulling the info from the DB.  But I see what you are saying about it still not working without the DB and just having the class.

Now, it may actually be stated that way in the docs, I just do not recall it as such. and if it is (or if you've updated it), then I certainly apologize.

**Comment by ikwattro** [comment]
@JoeThielen I understand about the docs. I added it in the new tutorial on read the docs. Writing docs is sometimes harder than writing code :)
